### PR TITLE
Implement the GraphQL Subscribe API

### PR DESCRIPTION
**Pull request by Ashish-Nanda**
*Description of changes:*
This is a draft PR for the GraphQL subscribe API. There is more polish required and some upcoming changes before this is complete:
- Replace the request `cancelToken` with an `id` or remove entirely.

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: packages/amplify_api/android/src/main/kotlin/com/amazonaws/amplify/amplify_api/FlutterApiErrorMessage.kt [hunk 1]
```diff
@@ -21,6 +21,8 @@ enum class FlutterApiErrorMessage {
 
     AMPLIFY_API_QUERY_FAILED,
     AMPLIFY_API_MUTATE_FAILED,
+    AMPLIFY_API_SUBSCRIBE_FAILED,
+    AMPLIFY_API_FAILED_TO_ESTABLISH_SUBSCRIPTION,
```
Comment (haverchuck) [inline packages/amplify_api/android/src/main/kotlin/com/amazonaws/amplify/amplify_api/FlutterApiErrorMessage.kt:25]: Aren't we removing these in favor of the new error handling design?

File: packages/amplify_api/android/src/main/kotlin/com/amazonaws/amplify/amplify_api/FlutterApiError.kt [hunk 2]
```diff
@@ -33,7 +33,8 @@ class FlutterApiError {
             handler.post { flutterResult.error("AmplifyException", msg, errorMap) }
         }
 
-        private fun createErrorMap(@NonNull error: Exception): Map<String, Any> {
+        @JvmStatic
+        fun createErrorMap(@NonNull error: Exception): Map<String, Any> {
```
Comment (haverchuck) [inline packages/amplify_api/android/src/main/kotlin/com/amazonaws/amplify/amplify_api/FlutterApiError.kt:37]: Not needed as part of the new error handling design.

File: packages/amplify_api/lib/method_channel_api.dart [hunk 3]
```diff
@@ -55,6 +58,61 @@ class AmplifyAPIMethodChannel extends AmplifyAPI {
...
+
+    //TODO: Either re-name the cancelToken field to id, or remove entirely
+    final subscriptionId = request.cancelToken;
+
+    _setupSubscription(
+      id: subscriptionId,
+      request: request,
+      onEstablished: onEstablished,
+    );
+
```
Comment (haverchuck) [inline packages/amplify_api/lib/method_channel_api.dart:81]: This *might* be a terrible idea - but *if* we want to return an actual StreamSubscription as a member of the operation class, we create a StreamController separate from the _subscription one, and then relay events from _subscription once all the appropriate parsing has been done.

I guess one issue here would be customers manipulating the returned StreamSubscription ('myStream', for example) and expecting cancelling it to have the same affect as cancelling the operation.

Comment (Amplifiyer) [reply packages/amplify_api/lib/method_channel_api.dart:81]: Ashish and I talked about this, we need to rearchitect this implementation to later support returning a StreamSubscription more natively. The way we are using streams are
1. Setup using methodChannel
2. Send data through eventChannel
3. Close/cancel using methodChannel

is not ideal where all of these steps can be done using just eventChannels. While calling `receiveBroadcaseStream([arguments])`, you pass your request params(such as GraphQLRequest) and flutter will establish a dedicated stream for you for this graphql subscription.

File: packages/amplify_api/test/amplify_api_subscribe_test.dart [hunk 4]
```diff
@@ -0,0 +1,105 @@
+/*
+ * Copyright 2020 Amazon.com, Inc. or its affiliates. All Rights Reserved.
```
Comment (Amplifiyer) [inline packages/amplify_api/test/amplify_api_subscribe_test.dart:2]: all new files should have 2021

File: packages/amplify_api_plugin_interface/lib/src/GraphQL/GraphQLSubscriptionOperation.dart [hunk 5]
```diff
@@ -13,14 +13,15 @@
  * permissions and limitations under the License.
  */
 
+import 'package:flutter/services.dart';
 import 'package:flutter/foundation.dart';
-import './types.dart';
 
-class GraphQLOperation<T> {
+const MethodChannel _channel = MethodChannel('com.amazonaws.amplify/api');
```
Comment (Amplifiyer) [inline packages/amplify_api_plugin_interface/lib/src/GraphQL/GraphQLSubscriptionOperation.dart:19]: Do we need this channel here?

File: packages/amplify_flutter/lib/categories/amplify_api_category.dart [hunk 6]
```diff
@@ -45,6 +45,20 @@ class APICategory {
...
+      Function() onEstablished,
+      Function(dynamic) onError,
+      Function() onDone}) {
+    return plugins[0].subscribe(
+        request: request,
+        onEstablished: onEstablished,
+        onData: onData,
+        onError: onError,
+        onDone: onDone);
+  }
```
Comment (Amplifiyer) [inline packages/amplify_flutter/lib/categories/amplify_api_category.dart:60]: To be consistent with other methods in this file, throw an exception if number of plugins != 1

File: packages/amplify_api/ios/Classes/FlutterGraphQLApi.swift [hunk 7]
```diff
@@ -131,4 +131,74 @@ class FlutterGraphQLApi {
...
+                                    }
+                                    if(established) {
+                                        let errorMap = FlutterApiError.createErrorMap(localizedError: apiError.errorDescription, recoverySuggestion: apiError.errorDescription)
+                                        let flutterError = FlutterError(
+                                            code: "AmplifyException",
+                                            message: FlutterApiErrorMessage.SUBSCRIBE_FAILED.rawValue,
+                                            details: errorMap)
+                                        graphQLSubscriptionsStreamHandler.sendError(flutterError: flutterError)
+                                    } else {
+                                        FlutterApiError.handleAPIError(flutterResult: flutterResult, error: apiError, msg: FlutterApiErrorMessage.ESTABLISH_SUBSCRIPTION_FAILED.rawValue)
```
[the remark this answers is not in the document]
Comment (Ashish-Nanda) [reply packages/amplify_api/ios/Classes/FlutterGraphQLApi.swift:189]: That will not happen because we are setting the `established` flag to `true` before calling `flutterResult` in the success case.
In the case of an exception, we call `flutterResult` only if `established` is still false, and the subscription connection did not establish itself. If the subscription is established, and exceptions are encountered after that, they are sent through the `graphQLSubscriptionsStreamHandler`

File: packages/amplify_api/ios/Classes/SwiftAmplifyApiPlugin.swift [hunk 8]
```diff
@@ -46,29 +51,30 @@ public class SwiftAmplifyApiPlugin: NSObject, FlutterPlugin {
...
+            case "put": FlutterRestApi.put(flutterResult: result, arguments: arguments, bridge: bridge)
+            case "delete": FlutterRestApi.delete(flutterResult: result, arguments: arguments, bridge: bridge)
+            case "head": FlutterRestApi.head(flutterResult: result, arguments: arguments, bridge: bridge)
+            case "patch": FlutterRestApi.patch(flutterResult: result, arguments: arguments, bridge: bridge)
+            case "query":
+                FlutterGraphQLApi.query(flutterResult: result, request: arguments, bridge: bridge)
+            case "mutate":
+                FlutterGraphQLApi.mutate(flutterResult: result, request: arguments, bridge: bridge)
+            case "subscribe":
+                FlutterGraphQLApi.subscribe(flutterResult: result, request: arguments, bridge: bridge, graphQLSubscriptionsStreamHandler:graphQLSubscriptionsStreamHandler)
```
[the remark this answers is not in the document]
Comment (Ashish-Nanda) [reply packages/amplify_api/ios/Classes/SwiftAmplifyApiPlugin.swift:75]: Done

File: packages/amplify_flutter/lib/categories/amplify_api_category.dart (hunk 6 above)
Comment (Ashish-Nanda) [reply packages/amplify_flutter/lib/categories/amplify_api_category.dart:60]: Done

File: packages/amplify_api/test/amplify_api_subscribe_test.dart (hunk 4 above)
Comment (Ashish-Nanda) [reply packages/amplify_api/test/amplify_api_subscribe_test.dart:2]: Done

File: packages/amplify_api/ios/Classes/GraphQLSubscriptionsStreamHandler.swift [hunk 9]
```diff
@@ -0,0 +1,42 @@
+/*
+ * Copyright 2020 Amazon.com, Inc. or its affiliates. All Rights Reserved.
```
[the remark this answers is not in the document]
Comment (Ashish-Nanda) [reply packages/amplify_api/ios/Classes/GraphQLSubscriptionsStreamHandler.swift:2]: Done

File: packages/amplify_api/ios/Classes/FlutterApiErrorMessage.swift [hunk 10]
```diff
@@ -19,6 +19,8 @@ enum FlutterApiErrorMessage: String {
     case MALFORMED = "AMPLIFY_REQUEST_MALFORMED"
     case QUERY_FAILED = "AMPLIFY_API_QUERY_FAILED"
     case MUTATE_FAILED = "AMPLIFY_API_MUTATE_FAILED"
+    case SUBSCRIBE_FAILED = "AMPLIFY_API_SUBSCRIBE_FAILED"
+    case ESTABLISH_SUBSCRIPTION_FAILED = "AMPLIFY_API_FAILED_TO_ESTABLISH_SUBSCRIPTION"
```
Comment (haverchuck) [inline packages/amplify_api/ios/Classes/FlutterApiErrorMessage.swift:23]: This change shouldn't be necessary, right?  If anything, we would be casting to a new exception type?

File: packages/amplify_api/android/src/main/kotlin/com/amazonaws/amplify/amplify_api/FlutterApiError.kt [hunk 11]
```diff
@@ -0,0 +1,58 @@
+/*
+ * Copyright 2020 Amazon.com, Inc. or its affiliates. All Rights Reserved.
```
Comment (Amplifiyer) [inline packages/amplify_api/android/src/main/kotlin/com/amazonaws/amplify/amplify_api/FlutterApiError.kt:2]: 2021

File: packages/amplify_api/android/src/main/kotlin/com/amazonaws/amplify/amplify_api/FlutterApiErrorMessage.kt [hunk 12]
```diff
@@ -0,0 +1,47 @@
...
+ * express or implied. See the License for the specific language governing
+ * permissions and limitations under the License.
+ */
+
+package com.amazonaws.amplify.amplify_api
+
+enum class FlutterApiErrorMessage {
+    AMPLIFY_REQUEST_MALFORMED,
+    ERROR_CASTING_INPUT_IN_PLATFORM_CODE,
+
```
Comment (Amplifiyer) [inline packages/amplify_api/android/src/main/kotlin/com/amazonaws/amplify/amplify_api/FlutterApiErrorMessage.kt:21]: Shouldn't this file be deleted?

File: packages/amplify_api/android/src/main/kotlin/com/amazonaws/amplify/amplify_api/FlutterApiError.kt [hunk 13]
```diff
@@ -0,0 +1,58 @@
...
+import android.os.Looper
+import androidx.annotation.NonNull
+import com.amplifyframework.api.ApiException
+import io.flutter.plugin.common.MethodChannel
+import io.flutter.plugin.common.MethodChannel.Result
+
+
+class FlutterApiError {
+    companion object {
+        private val handler = Handler(Looper.getMainLooper())
```
Comment (Amplifiyer) [inline packages/amplify_api/android/src/main/kotlin/com/amazonaws/amplify/amplify_api/FlutterApiError.kt:28]: Shouldn't this file be deleted?

File: packages/amplify_api/android/src/test/kotlin/com/amazonaws/amplify/amplify_api/GraphQLApiUnitTests.kt [hunk 14]
```diff
@@ -296,6 +299,140 @@ class GraphQLApiUnitTests {
...
+                mockResult
+        )
+
+        verify(mockResult).error(
+                "ApiException",
+                ExceptionMessages.defaultFallbackExceptionMessage,
+                mapOf(
+                        "message" to ExceptionMessages.missingExceptionMessage,
+                        "recoverySuggestion" to ExceptionMessages.missingRecoverySuggestion,
+                        "underlyingException" to "AmplifyException{message=The graphQL document request argument was not passed as a String, cause=kotlin.TypeCastException: null cannot be cast to non-null type kotlin.String, recoverySuggestion=The request should include the graphQL document as a String}"
```
Comment (Amplifiyer) [inline packages/amplify_api/android/src/test/kotlin/com/amazonaws/amplify/amplify_api/GraphQLApiUnitTests.kt:375]: nit: wrap big lines.

File: packages/amplify_api/android/src/test/kotlin/com/amazonaws/amplify/amplify_api/GraphQLApiUnitTests.kt [hunk 15]
```diff
@@ -296,6 +299,140 @@ class GraphQLApiUnitTests {
...
+                )
+        )
+    }
+
+    @Test
+    fun test_subscribe_api_error() {
```
Comment (Amplifiyer) [inline packages/amplify_api/android/src/test/kotlin/com/amazonaws/amplify/amplify_api/GraphQLApiUnitTests.kt:381]: I believe `sendError` scenario is not tested, i.e. when subscription is established and then error is received (An error scenario more likely to happen)

Same for iOS.

File: packages/amplify_api/ios/Classes/FlutterApiErrorMessage.swift [hunk 16]
```diff
@@ -0,0 +1,44 @@
...
+ *
+ * or in the "license" file accompanying this file. This file is distributed
+ * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+ * express or implied. See the License for the specific language governing
+ * permissions and limitations under the License.
+ */
+
+import Foundation
+enum FlutterApiErrorMessage: String {
+    case CASTING = "ERROR_CASTING_INPUT_IN_PLATFORM_CODE"
```
Comment (Amplifiyer) [inline packages/amplify_api/ios/Classes/FlutterApiErrorMessage.swift:18]: Shouldn't these files be deleted?

File: packages/amplify_api/ios/Classes/FlutterApiResponse.swift [hunk 17]
```diff
@@ -46,5 +46,32 @@ class FlutterApiResponse {
                                                 details: ErrorUtil.createSerializedError(message: errorDescription, recoverySuggestion: recoverySuggestion, underlyingError: nil))
         }
     }
+    
+    static func handleGraphQLErrorResponseEvent(graphQLSubscriptionsStreamHandler: GraphQLSubscriptionsStreamHandler, id: String, errorResponse: GraphQLResponseError<String>, failureMessage: String) {
```
Comment (Amplifiyer) [inline packages/amplify_api/ios/Classes/FlutterApiResponse.swift:50]: wrap big lines.

File: packages/amplify_api/ios/Classes/FlutterApiResponse.swift [hunk 18]
```diff
@@ -46,5 +46,32 @@ class FlutterApiResponse {
...
+            print("Received a partially successful GraphQL subscription event: \(payload)")
+            graphQLSubscriptionsStreamHandler.sendEvent(payload: payload, id: id, type: GraphQLSubscriptionEventTypes.DATA)
+        case .transformationError(let rawResponse, let error):
+            print("Received a partially successful GraphQL subscription event with a transformation error: \(error)")
+            let details = FlutterApiErrorHandler.createSerializedError(error: error)
+            graphQLSubscriptionsStreamHandler.sendError(errorCode: "ApiException", details: details)
+        case .unknown(let errorDescription, let recoverySuggestion, let error):
+            print("An unknown error occured: \(errorDescription)")
+            let details = FlutterApiErrorHandler.createSerializedError(error: APIError.unknown(errorDescription, recoverySuggestion, error))
+            graphQLSubscriptionsStreamHandler.sendError(errorCode: "ApiException", details: details)
```
Comment (Amplifiyer) [inline packages/amplify_api/ios/Classes/FlutterApiResponse.swift:73]: These cases didn't seem to be tested.

File: packages/amplify_api/example/ios/unit_tests/GraphQLApiUnitTests.swift [hunk 19]
```diff
@@ -226,9 +228,151 @@ class GraphQLApiUnitTests: XCTestCase {
...
+        
+        FlutterGraphQLApi.subscribe(
+            flutterResult: { (result) -> Void in
+                if let exception = result as? FlutterError {
+                    XCTAssertEqual("AmplifyException", exception.code)
+                    XCTAssertEqual(FlutterApiErrorMessage.MALFORMED.rawValue, exception.message)
+                    let errorMap: [String: Any] = exception.details as! [String : Any]
+                    XCTAssertEqual(
+                        ["platform": "iOS", "localizedErrorMessage": "The graphQL document request argument was not passed as a String", "recoverySuggestion": "The request should include the graphQL document as a String"],
+                        errorMap["PLATFORM_EXCEPTIONS"] as? [String: String])
```
Comment (Amplifiyer) [inline packages/amplify_api/example/ios/unit_tests/GraphQLApiUnitTests.swift:357]: Isn't this wrong?

File: packages/amplify_api/example/ios/unit_tests/GraphQLApiUnitTests.swift [hunk 20]
```diff
@@ -226,9 +228,151 @@ class GraphQLApiUnitTests: XCTestCase {
...
+        
+        FlutterGraphQLApi.subscribe(
+            flutterResult: { (result) -> Void in
+                if let exception = result as? FlutterError {
+                    XCTAssertEqual("AmplifyException", exception.code)
+                    XCTAssertEqual(FlutterApiErrorMessage.ESTABLISH_SUBSCRIPTION_FAILED.rawValue, exception.message)
+                    let errorMap: [String: Any] = exception.details as! [String : Any]
+                    XCTAssertEqual(
+                        ["platform": "iOS", "localizedErrorMessage": "test error", "recoverySuggestion": "test recovery suggestion"],
+                        errorMap["PLATFORM_EXCEPTIONS"] as? [String: String])
```
Comment (Amplifiyer) [inline packages/amplify_api/example/ios/unit_tests/GraphQLApiUnitTests.swift:312]: Isn't this wrong?

File: packages/amplify_api/ios/Classes/SwiftAmplifyApiPlugin.swift [hunk 21]
```diff
@@ -19,19 +19,24 @@ import Amplify
 import AmplifyPlugins
 
 public class SwiftAmplifyApiPlugin: NSObject, FlutterPlugin {
-
+
```
Comment (Amplifiyer) [inline packages/amplify_api/ios/Classes/SwiftAmplifyApiPlugin.swift:22]: This file got a lot of whitespace changes

File: packages/amplify_api/lib/method_channel_api.dart [hunk 22]
```diff
@@ -56,6 +59,60 @@ class AmplifyAPIMethodChannel extends AmplifyAPI {
...
+        final errors = _deserializeGraphQLResponseErrors(payload);
+
+        GraphQLResponse<T> response =
+            GraphQLResponse<T>(data: payload['data'], errors: errors);
+        onData(response);
+      }
+    });
+
+    //TODO: This should convert the exception to APIException if present
+    _subscription.onError(onError);
```
Comment (Amplifiyer) [inline packages/amplify_api/lib/method_channel_api.dart:106]: Are we not doing this?

File: packages/amplify_api/lib/method_channel_api.dart [hunk 23]
```diff
@@ -179,8 +255,7 @@ class AmplifyAPIMethodChannel extends AmplifyAPI {
   }
 
   //TODO: Deserialize all fields of the GraphQLResponseError as per spec
-  List<GraphQLResponseError> _deserializeGraphQLResponseErrors(
-      Map<String, dynamic> response) {
+  List<GraphQLResponseError> _deserializeGraphQLResponseErrors(Map response) {
```
Comment (Amplifiyer) [inline packages/amplify_api/lib/method_channel_api.dart:258]: Is this change needed?

File: packages/amplify_api/test/amplify_api_subscribe_test.dart [hunk 24]
```diff
@@ -0,0 +1,105 @@
...
+
+  handler(event) {
+    ServicesBinding.instance.defaultBinaryMessenger.handlePlatformMessage(
+      channelName,
+      event,
+      (ByteData reply) {},
+    );
+  }
+
+  test('Subscribe API sets up the subscription over the method channel',
```
Comment (Amplifiyer) [inline packages/amplify_api/test/amplify_api_subscribe_test.dart:42]: Tests for receiving errors before/after onEstablish seem to be missing.

File: packages/amplify_api/example/ios/unit_tests/GraphQLApiUnitTests.swift (hunk 20 above)
Comment (Ashish-Nanda) [reply packages/amplify_api/example/ios/unit_tests/GraphQLApiUnitTests.swift]: The tests were correct as per the original error handling approach. I've updated them now after the refactor.

File: packages/amplify_api/example/ios/unit_tests/GraphQLApiUnitTests.swift (hunk 19 above)
Comment (Ashish-Nanda) [reply packages/amplify_api/example/ios/unit_tests/GraphQLApiUnitTests.swift]: The tests were correct as per the original error handling approach. I've updated them now after the refactor.

File: packages/amplify_api_plugin_interface/lib/src/GraphQL/GraphQLSubscriptionOperation.dart (hunk 5 above)
Comment (Ashish-Nanda) [reply packages/amplify_api_plugin_interface/lib/src/GraphQL/GraphQLSubscriptionOperation.dart]: Removed

File: packages/amplify_api/ios/Classes/FlutterApiErrorMessage.swift (hunk 10 above)
Comment (Ashish-Nanda) [reply packages/amplify_api/ios/Classes/FlutterApiErrorMessage.swift]: This change was based on the original error handling approach. I've now updated the code based on the error handling refactor.

File: packages/amplify_api/lib/method_channel_api.dart (hunk 23 above)
Comment (Ashish-Nanda) [reply packages/amplify_api/lib/method_channel_api.dart:258]: Reverted

File: packages/amplify_api/android/src/test/kotlin/com/amazonaws/amplify/amplify_api/GraphQLApiUnitTests.kt (hunk 14 above)
Comment (Ashish-Nanda) [reply packages/amplify_api/android/src/test/kotlin/com/amazonaws/amplify/amplify_api/GraphQLApiUnitTests.kt:375]: Done. Also moved to a constant.

File: packages/amplify_api/ios/Classes/SwiftAmplifyApiPlugin.swift (hunk 21 above)
Comment (Ashish-Nanda) [reply packages/amplify_api/ios/Classes/SwiftAmplifyApiPlugin.swift:22]: I reformatted it using Xcode

File: packages/amplify_api/ios/Classes/FlutterApiResponse.swift (hunk 17 above)
Comment (Ashish-Nanda) [reply packages/amplify_api/ios/Classes/FlutterApiResponse.swift]: Done

File: packages/amplify_api/lib/method_channel_api.dart (hunk 22 above)
Comment (Ashish-Nanda) [reply packages/amplify_api/lib/method_channel_api.dart:106]: Testing this, will update soon.

File: packages/amplify_api/ios/Classes/SwiftAmplifyApiPlugin.swift (hunk 21 above)
Comment (Amplifiyer) [reply packages/amplify_api/ios/Classes/SwiftAmplifyApiPlugin.swift:22]: This can be easily fixed: Xcode->Preferences->Text Editing->Automatically trim trailing whitespace + Including whitespace-only lines and then reformat.

File: packages/amplify_api/ios/Classes/GraphQLSubscriptionsStreamHandler.swift [hunk 25]
```diff
@@ -0,0 +1,42 @@
...
+        self.eventSink = events
+        return nil
+    }
+
+    func sendEvent(payload: [String : Any]?, id: String, type: GraphQLSubscriptionEventTypes) {
+        let result: [String: Any?] = [
+            "id": id,
+            "type": type.rawValue,
+            "payload": payload
+        ]
```
[the remark this answers is not in the document]
Comment (Ashish-Nanda) [reply packages/amplify_api/ios/Classes/GraphQLSubscriptionsStreamHandler.swift:32]: Done

**Review by Amplifiyer — APPROVED** [review]
Looks good!

File: packages/amplify_api/android/src/test/kotlin/com/amazonaws/amplify/amplify_api/GraphQLApiUnitTests.kt (hunk 15 above)
Comment (Ashish-Nanda) [reply packages/amplify_api/android/src/test/kotlin/com/amazonaws/amplify/amplify_api/GraphQLApiUnitTests.kt:495]: Done for both iOS and Android.
